# **Computational Data Analysis with Python - Week 2** 🔍 📊 🐍

## 📌 Overview
This second week tutorial closes the introductory part to python coding. You will go through these topics: libraries, functions, OOP and files managing. Each part contains exercises you must solve to understand deeply the concepts and then applying them to the Hackathon.


### 🧠 **Learning objectives**
- Importing basic Libraries and exploting their functionalities
- Understanding the basic concepts of functions and applying them to improve readability of the code
- Using Gemini conciously, how to prompt and how to debug improper prompting

### 🐍 **New in Python**
- pip
- Libraries
- Functions
- Classes and OOP
- Files

### 1. Libraries
- **1.1. Library**
- **1.2. pip - Python Package Manager**

### 2. Functions
- **2.1. Syntax**
- **2.2. High-order function**
- **2.3. Anonymous function**

### 3. Classes and OOP

### 4. Files

### 5. Hackaton: the grocery warehouse Pt.2📦


## 1. **Libraries**
Some very smart people built lower level pieces of code that High-IQ people can exploit for their own porpuses (This is us), and remember: "I choose a lazy person to do a hard job. Because a lazy person will find an easy way to do it."
— **Attributed to Bill Gates**

So, let's trust the scientific research and go on with our tutorial.


**1.1. Library**: is a collection of non-volatile resources used by computer programs, often to manage data or perform complex calculations.
One useful example could be the library **math** that improve the readability of formula, moreover, it has preconfigured constants: π and e. It is part of python standard library, to exploit its features you must only move from disk to RAM by means `import`.

In [ ]:
# To use a library you have to "import" it from Disk memory to RAM
import math

# Let's calculate precisely a Area of a circle with R=1:
R = 1
Area = math.pi * R**2
print(f"The Area of the circle is: {Area}")

# Let's evaluate the square root of 16:
print(f"The square root of 16 is: {math.sqrt(16)}")

In [ ]:
# To save memory is also possible to import in the RAM only a module:
from math import sqrt

print(sqrt(25))         # 5.0
# print(pi)             # This would fail because we ONLY took sqrt!

In [ ]:
# Moreover, to save time you can use the short name:
import math as m

print(m.factorial(5))   # 120 (5*4*3*2*1)

**1.2. pip - Python Package Manager**
pip is a standalone program that serves as a package manager to install python libraries. Practically speaking, if import move the library from disk to RAM, `pip install` move it from internet to the disk. Here below will find summarized some useful py libraries that need to be installed:  

In [ ]:
# STEP 1: INSTALLATION (Bringing the tool home)
# Run this only once in your notebook
!pip install emoji

# STEP 2: IMPORTING (Putting the tool on the workbench)
import emoji

# STEP 3: USAGE
print(emoji.emojize("Python is fun! :snake: :rocket:"))

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 608.4/608.4 kB 9.2 MB/s eta 0:00:00
Python is fun! 🐍 🚀


In [ ]:
# Faker: It can generate names, addresses, emails, and even fake profiles in seconds
!pip install Faker
from faker import Faker

fake = Faker()

# Create 5 fake customer profiles
for _ in range(5):
    print(f"Name: {fake.name()} | Job: {fake.job()}")

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.0/2.0 MB 21.7 MB/s eta 0:00:00
Name: Robert Johnson | Job: Restaurant manager
Name: Daniel Mills | Job: Futures trader
Name: Cynthia Walker | Job: Insurance risk surveyor
Name: Benjamin Mccall | Job: Production assistant, television
Name: Lisa Smith | Job: Make


In [ ]:
# tqdm: Have you ever run a loop that takes 30 seconds and wondered if Colab crashed? tqdm (derived from the Arabic word taqadum, meaning "progress") adds a visual progress bar to your loops.
!pip install tqdm
from tqdm import tqdm
import time

# Simulate a long audit of 100 items
for item in tqdm(range(100)):
    time.sleep(0.05) # Simulate "work" happening

100%|██████████| 100/100 [00:05<00:00, 19.75it/s]


In [ ]:
# Rich: to enrich your plotting
!pip install rich
from rich import print
from rich.table import Table

table = Table(title="Grocery Store Audit")

table.add_column("Item", style="cyan")
table.add_column("Price", style="green")

table.add_row("Apple", "$1.50")
table.add_row("Bread", "$2.20")

print(table)

  Grocery Store  
      Audit      
┏━━━━━━━┳━━━━━━━┓
┃ Item  ┃ Price ┃
┡━━━━━━━╇━━━━━━━┩
│ Apple │ $1.50 │
│ Bread │ $2.20 │
└───────┴───────┘

## 2. **Functions**
A function is a black-box part of the code used to improve **readability** of the main/backbone code.
When python hits a functiob call, it doesn't just read the next line. It performs a Control Flow Jump: it "pauses" the main script, it "jumps" to the function block, it creates a Local Scope (a temporary memory space), runs the code and ,eventually, goes back to the exact spot left carrying the return value lika a courier.
Depending on the complexity of the code, you can define the function before in the same source file and directly use it, or you can define it ina different source file and **import** it.

**2.1. Syntax**
``` python
def myFunction(list_of_input):      # Order matters
---> """
     Block note usually exploited to explain the features of the
     function:
     - input:....
     - output:... Force [kN]..
     """
     myVariable = 1   # Valid only inside the function
     return list_of_output
```

In [ ]:
# A function to retrieve the max in between two numbers:

def myMax(a,b):
   if a>b:
     return a
   else:
     return b
a = 50
b = 5
print(myMax(a,b))

50

In [ ]:
# If I have to output more than one variables:
def toUpperLower(input_string):
    upper_string = input_string.upper()
    lower_string = input_string.lower()
    return upper_string, lower_string

myString = 'CiaoCIAOciao'
upper, lower = toUpperLower(myString)
print(upper)
print(lower)
uplow = toUpperLower(myString)
print(uplow)  # In this case I will obtain a tuple (Not changeble item)

CIAOCIAOCIAO

ciaociaociao

('CIAOCIAOCIAO', 'ciaociaociao')

In [ ]:
# Parameters of a function:
def myGreater(a,b=0):       # b=0 as default value, if nothing is specified
   return a>b
result = myGreater(-1)
print(result)

# It is possible to reassign the value of b:
result = myGreater(-1,-10)
print(result)

False
True


In [ ]:
# Named parameters:
def myFunction(a,b,c,d):
    print(a)
    print(b)
    print(c)
    print(d)
    return a+b+c+d

sum = myFunction(d=1,a=2,c=3,b=4)  # The order doesn't matter if you explicitely call the input variables
print(sum)

sum_2 = myFunction(1,2,3,4)        # In this case the value are assigned order-wise
print(sum_2)

d=4
sum_3 = myFunction(a=1, b=2, c=3, d=d)
print(sum_3)

2
4
3
1
10
1
2
3
4
10
1
2
3
4
10


**2.2. High-order function**
In the previous examples we gave variables as input to a function and retrive output variables. If the input and output of a functions are functions themself, this is called **high-order** computing. In the following, it is presented `map` and its syntax. This command will apply a function to all the elements of a list (It is actually better than a for loop because computation run in parallel). \
N.B. It is better than a for-loop beacuse it runs functions in paralell.

In [ ]:
# map(function, iterable,...) indicate the list of input to the functions
def toUpper(input_string):
    return input_string.upper()

myList = ['Halo','CIao']
upperList = map(toUpper,myList)
print(upperList)
print(list(upperList))

['HALO', 'CIAO']


**2.3. Anonymous function**
It is called `lambda`: deep inside python threat it as a function and it returns a function object. Moreover, it is very useful for simple one-liner function (it avoids to define a function from scratch) and whose result is immediately returned. \
``` python
result = lambda parameters:expression
```


In [ ]:
# Convert a unit of measure:

diameter_cm = 20
diameter_inch = lambda x: x/2.54
print(f"The lambda fun. is useful to convert a measure from {diameter_cm} cm to {diameter_inch(diameter_cm)} inches")

The lambda fun. is useful to convert a measure from 20cm to 7.874015748031496inches


In [ ]:
# Convert a list of measures:

diameter_cm = [20,30,40]
diameter_inch = list(map(lambda x: x/2.54, diameter_cm))   # Need to declare as a list()
print(diameter_inch)

[7.874015748031496, 11.811023622047244, 15.748031496062993]


**EX.1 Write a function that performs safety check for a rebar subjected to tensile forces and output the :**
```python
diameter = 10          # [cm]
safety_coeff = 1.15    # [-]
force = 1000           # [N]
E = 2e5                # [MPa]
```

**EX.2 Do the smae for this list of rebars**
```python
diameters_inch = [7.5, 8.2, 11.3]  # [inch]
safety_coeff = 1.15    # [-]
force = 2000           # [N]
E = 2e5                # [MPa]
```

## 3. **Classes and Object-oriented Programming (OOP)**

**Keep in mind:**"An Object is an instance of a Class"
ex. BMW 320 is an object from the class Auto

Almost everything is an object at a lower level. It is not strictly necessary to apply this knowledge but it is warmly recommended to understand it.

```python
class TensileBar:
    # 1. THE CONSTRUCTOR (Identity & State)
    def __init__(self, E, phi):
        self.E = E                  # Attribute: steel Young Modulus in MPa
        self.phi = phi              # Attribute: Cross-section diameter in mm
        self.force = 0              # Initial state: No load applied yet

    # 2. THE METHODS (Behavior)
    def apply_load(self, normal_force):
        """Changes the state of the object."""
        self.force_n = normal_force
        print(f"Applied {newtons}N to {self.name}")

    def get_area(self):
        """Calculates internal stress based on current state."""
        import math
        area = math.pi * ((self.diameter/2)**2)
        return area

     def get_resistance(self,area):
        """Calculates internal stress based on current state."""
        resistance = area * self.E
        return resistance
```


**EX.3 Check with Gemini what is the "Inheritance" of a Class. Create a Class called 'Point2D' taking as input and attributes x and y coordinate of a point in Cartesian plane. Define a method called 'movePoint' that move P in the plane of dx and dy. Eventually use inheritance to call the class 'Point3D' that manages a point in 3D, updating as well the function movePoint.**

## 4. **Files**
Create, manage and read a .txt file!
- Create a .txt file in your desktop by copy-pasting:
```
    Test_ID: 001 | Status: Pending
    Test_ID: 002 | Status: Pending
    Test_ID: 003 | Status: Pending
```
- drag in the colab (folder icon on the left-handed side)
- Alterantively, run the code below

In [ ]:
# STEP 1: Creating the file
rows = [
    "Test_ID: 001 | Status: Pending \n",
    "Test_ID: 002 | Status: Pending \n",
    "Test_ID: 003 | Status: Pending \n"
]

with open("lab_log.txt", "w") as file:
    file.writelines(rows)

print("File created successfully.")

File created successfully.


In [ ]:
# STEP 2: Modifying a specific line
# 1. READ the current state
with open("lab_log.txt", "r") as file:
    data = file.readlines()

# 2. MODIFY the second item (index 1) in our list
# We replace the whole string for that row
data[1] = data[1].replace("Pending", "PASSED")

# 3. WRITE it back (Overwrite mode 'w')
with open("lab_log.txt", "w") as file:
    file.writelines(data)

print("Modification complete: Row 2 updated.")

Modification complete: Row 2 updated.


In [ ]:
# STEP 3: Adding a fourth row
new_entry = "Test_ID: 004 | Status: Pending \n"

with open("lab_log.txt", "a") as file:
    file.write(new_entry)

print("New row appended.")

New row appended.


**EX.4 Check with Gemini the difference in between a .csv and .tsv file.**

## 5. **HACKATON: the grocery warehouse Pt.2** 📦
A little local grocery store hires you as accountant. They provide to you a list of item in the warehouse and you are in charge of retrievieing data from the register and assist the owner to restock, calculate tax and other stuffs...
Luckly, you attended the Computational Data Analysis with Python course and you solve all the tasks of the week in 30 Minutes!!

In [ ]:
import kagglehub
path = kagglehub.dataset_download("chinmayshanbhag/big-basket-products")

100%|██████████| 525k/525k [00:00<00:00, 83.9MB/s]

Extracting files...


In [ ]:
import os
import pandas as pd
# Database management DO NOT TOUCH THIS!
# List the contents of the downloaded dataset directory to find the CSV file
# The 'path' variable is available from the previous step
dataset_files = os.listdir(path)
csv_files = [f for f in dataset_files if f.endswith('.csv')]

if not csv_files:
    print(f"No CSV files found in the directory: {path}")
    csv_file_path = None
elif len(csv_files) == 1:
    csv_file_name = csv_files[0]
    csv_file_path = os.path.join(path, csv_file_name)
    print(f"Found CSV file: {csv_file_path}")
else:
    print(f"Multiple CSV files found: {csv_files}. Attempting to use the first one: {csv_files[0]}")
    csv_file_name = csv_files[0]
    csv_file_path = os.path.join(path, csv_file_name)

if csv_file_path:
    # Load the CSV file into a pandas DataFrame
    df = pd.read_csv(csv_file_path)

    print("\nFirst 5 rows of the DataFrame:")
    display(df.head(20))

    print("\nLast 5 rows of the DataFrame:")
    display(df.tail())
else:
    print("Could not proceed as no CSV file was found.")

Found CSV file: /root/.cache/kagglehub/datasets/chinmayshanbhag/big-basket-products/versions/1/BigBasket.csv

First 5 rows of the DataFrame:


,ProductName,Brand,Price,DiscountPrice,Image_Url,Quantity,Category,SubCategory,Absolute_Url
0,Onion (Loose),Fresho,69.75,52.00,https://www.bigbasket.com/media/uploads/p/l/40...,2 kg,Fruits & Vegetables,"Potato, Onion & Tomato",https://www.bigbasket.com/pd/40075537/fresho-o...
1,Onion,Fresho,174.35,130.00,https://www.bigbasket.com/media/uploads/p/l/12...,5 kg,Fruits & Vegetables,"Potato, Onion & Tomato",https://www.bigbasket.com/pd/1201414/fresho-on...
2,Onion (Loose),Fresho,34.87,26.00,https://www.bigbasket.com/media/uploads/p/l/10...,1 kg,Fruits & Vegetables,"Potato, Onion & Tomato",https://www.bigbasket.com/pd/10000148/fresho-o...
3,Onion,Fresho,69.74,52.00,https://www.bigbasket.com/media/uploads/p/l/12...,2 kg,Fruits & Vegetables,"Potato, Onion & Tomato",https://www.bigbasket.com/pd/1201413/fresho-on...
4,Onion (Loose),Fresho,174.37,129.00,https://www.bigbasket.com/media/uploads/p/l/10...,5 kg,Fruits & Vegetables,"Potato, Onion & Tomato",https://www.bigbasket.com/pd/10000150/fresho-o...
5,"Farm Eggs - Table Tray, Medium, Antibiotic Res...",Fresho,110.00,78.00,https://www.bigbasket.com/media/uploads/p/l/40...,12 pcs,"Eggs, Meat & Fish",Farm Eggs,https://www.bigbasket.com/pd/40072320/fresho-f...
6,"Farm Eggs - Table Tray, Medium, Antibiotic Res...",Fresho,210.00,181.00,https://www.bigbasket.com/media/uploads/p/l/15...,30 pcs,"Eggs, Meat & Fish",Farm Eggs,https://www.bigbasket.com/pd/150502/fresho-far...
7,Potato (Loose),Fresho,50.32,38.75,https://www.bigbasket.com/media/uploads/p/l/40...,1 kg,Fruits & Vegetables,"Potato, Onion & Tomato",https://www.bigbasket.com/pd/40048457/fresho-p...
8,Potato,Fresho,100.65,77.50,https://www.bigbasket.com/media/uploads/p/l/40...,2 kg,Fruits & Vegetables,"Potato, Onion & Tomato",https://www.bigbasket.com/pd/40162469/fresho-p...
9,Eggs - Regular,Fresho,250.00,250.00,https://www.bigbasket.com/media/uploads/p/l/12...,5x6 pcs,"Eggs, Meat & Fish",Farm Eggs,https://www.bigbasket.com/pd/1205938/fresho-eg...



Last 5 rows of the DataFrame:


,ProductName,Brand,Price,DiscountPrice,Image_Url,Quantity,Category,SubCategory,Absolute_Url
8203,"Foil Alphabet Balloon - Letter V, Party Decor,...",Bvishal,49.0,25.0,https://www.bigbasket.com/media/uploads/p/l/40...,1 pc,Cleaning & Household,"Caps, Balloons & Candles",https://www.bigbasket.com/pd/40237020/bvishal-...
8204,"Decoration Balls - Used For Parties, Unique De...",Bvishal,399.0,249.0,https://www.bigbasket.com/media/uploads/p/l/40...,6 pcs,Cleaning & Household,Decorations,https://www.bigbasket.com/pd/40237682/bvishal-...
8205,Foundation Blending Brush - 230,PAC,465.0,232.5,https://www.bigbasket.com/media/uploads/p/l/40...,1 pc,Beauty & Hygiene,Makeup Accessories,https://www.bigbasket.com/pd/40175488/pac-foun...
8206,One Week Immunity Booster,The Healthy Company,799.0,799.0,https://www.bigbasket.com/media/uploads/p/l/40...,28 g,Beauty & Hygiene,Supplements & Proteins,https://www.bigbasket.com/pd/40200885/the-heal...
8207,"Blush & Glow - Green Apple Apricot Gel Scrub, ...",Lakme,250.0,175.0,https://www.bigbasket.com/media/uploads/p/l/40...,100 g,Beauty & Hygiene,Face Care,https://www.bigbasket.com/pd/40224773/lakme-bl...


In [ ]:
required_columns = ['ProductName', 'Price', 'DiscountPrice', 'Category']
df_filtered = df[required_columns].copy()

# Ensure prices are floats
df_filtered['Price'] = df_filtered['Price'].astype(float)

# Split 'Quantity' into 'qty' and 'unit_of_measure'
df_filtered[['qty', 'unit_of_measure']] = df['Quantity'].str.extract(r'(\d+\.?\d*)\s*([a-zA-Z]+)?')

# Convert 'qty' to numeric (float) and fill any NaN with 0
df_filtered['qty'] = pd.to_numeric(df_filtered['qty'], errors='coerce').fillna(0)

# Handle cases where unit might be missing (e.g., if only a number is present)
df_filtered['unit_of_measure'] = df_filtered['unit_of_measure'].fillna('')


display(df_filtered.head())
display(df_filtered.tail())


,ProductName,Price,DiscountPrice,Category,qty,unit_of_measure
0,Onion (Loose),69.75,52.0,Fruits & Vegetables,2.0,kg
1,Onion,174.35,130.0,Fruits & Vegetables,5.0,kg
2,Onion (Loose),34.87,26.0,Fruits & Vegetables,1.0,kg
3,Onion,69.74,52.0,Fruits & Vegetables,2.0,kg
4,Onion (Loose),174.37,129.0,Fruits & Vegetables,5.0,kg


,ProductName,Price,DiscountPrice,Category,qty,unit_of_measure
8203,"Foil Alphabet Balloon - Letter V, Party Decor,...",49.0,25.0,Cleaning & Household,1.0,pc
8204,"Decoration Balls - Used For Parties, Unique De...",399.0,249.0,Cleaning & Household,6.0,pcs
8205,Foundation Blending Brush - 230,465.0,232.5,Beauty & Hygiene,1.0,pc
8206,One Week Immunity Booster,799.0,799.0,Beauty & Hygiene,28.0,g
8207,"Blush & Glow - Green Apple Apricot Gel Scrub, ...",250.0,175.0,Beauty & Hygiene,100.0,g


In [ ]:
# Convert the DataFrame to a list of dictionaries
df_as_dict = df_filtered.to_dict(orient='records')

# Display the first 5 entries of the dictionary list to verify
print("First 5 rows of df_filtered as a list of dictionaries:")
for i in range(min(5, len(df_as_dict))):
    print(df_as_dict[i])

First 5 rows of df_filtered as a list of dictionaries:
{'ProductName': 'Onion (Loose)', 'Price': 69.75, 'DiscountPrice': 52.0, 'Category': 'Fruits & Vegetables', 'qty': 2.0, 'unit_of_measure': 'kg'}
{'ProductName': 'Onion', 'Price': 174.35, 'DiscountPrice': 130.0, 'Category': 'Fruits & Vegetables', 'qty': 5.0, 'unit_of_measure': 'kg'}
{'ProductName': 'Onion (Loose)', 'Price': 34.87, 'DiscountPrice': 26.0, 'Category': 'Fruits & Vegetables', 'qty': 1.0, 'unit_of_measure': 'kg'}
{'ProductName': 'Onion', 'Price': 69.74, 'DiscountPrice': 52.0, 'Category': 'Fruits & Vegetables', 'qty': 2.0, 'unit_of_measure': 'kg'}
{'ProductName': 'Onion (Loose)', 'Price': 174.37, 'DiscountPrice': 129.0, 'Category': 'Fruits & Vegetables', 'qty': 5.0, 'unit_of_measure': 'kg'}


### **Tasks with `df_as_dict`:**

1.  **Categorize items by price using a custom function.**
2.  **Calculate the percentage discount for each item using a lambda function.**
3.  **Create a list of formatted product descriptions using `map` with a lambda function.**

#### **Task 1: Categorize items by price using a custom function**

We'll define a function that takes an item (dictionary) and returns a price category ('Budget', 'Standard', 'Premium') based on its `DiscountPrice`.

In [ ]:
def get_price_category(item):
    price = item['DiscountPrice']
    if price < 50:
        return 'Budget'
    elif 50 <= price < 150:
        return 'Standard'
    else:
        return 'Premium'

# Apply the custom function to each item
item_price_categories = []
for item in df_as_dict:
    item_price_categories.append({
        'ProductName': item['ProductName'],
        'Category': item['Category'],
        'DiscountPrice': item['DiscountPrice'],
        'PriceCategory': get_price_category(item)
    })

print("First 5 items with price categories:")
for i in range(min(5, len(item_price_categories))):
    print(item_price_categories[i])

First 5 items with price categories:
{'ProductName': 'Onion (Loose)', 'Category': 'Fruits & Vegetables', 'DiscountPrice': 52.0, 'PriceCategory': 'Standard'}
{'ProductName': 'Onion', 'Category': 'Fruits & Vegetables', 'DiscountPrice': 130.0, 'PriceCategory': 'Standard'}
{'ProductName': 'Onion (Loose)', 'Category': 'Fruits & Vegetables', 'DiscountPrice': 26.0, 'PriceCategory': 'Budget'}
{'ProductName': 'Onion', 'Category': 'Fruits & Vegetables', 'DiscountPrice': 52.0, 'PriceCategory': 'Standard'}
{'ProductName': 'Onion (Loose)', 'Category': 'Fruits & Vegetables', 'DiscountPrice': 129.0, 'PriceCategory': 'Standard'}


#### **Task 2: Calculate the percentage discount for each item using a lambda function**

Here, we use a `lambda` function for a concise way to compute the percentage discount for each product. We'll handle cases where the original price might be zero to avoid division by zero errors.

In [ ]:
# Use a lambda function to calculate percentage discount
percentage_discounts = []
for item in df_as_dict:
    price = item['Price']
    discount_price = item['DiscountPrice']
    discount_percentage = (lambda p, dp: ((p - dp) / p * 100) if p > 0 else 0)(price, discount_price)
    percentage_discounts.append({
        'ProductName': item['ProductName'],
        'OriginalPrice': price,
        'DiscountPrice': discount_price,
        'DiscountPercentage': f"{discount_percentage:.2f}%"
    })

print("First 5 items with percentage discounts:")
for i in range(min(5, len(percentage_discounts))):
    print(percentage_discounts[i])

First 5 items with percentage discounts:
{'ProductName': 'Onion (Loose)', 'OriginalPrice': 69.75, 'DiscountPrice': 52.0, 'DiscountPercentage': '25.45%'}
{'ProductName': 'Onion', 'OriginalPrice': 174.35, 'DiscountPrice': 130.0, 'DiscountPercentage': '25.44%'}
{'ProductName': 'Onion (Loose)', 'OriginalPrice': 34.87, 'DiscountPrice': 26.0, 'DiscountPercentage': '25.44%'}
{'ProductName': 'Onion', 'OriginalPrice': 69.74, 'DiscountPrice': 52.0, 'DiscountPercentage': '25.44%'}
{'ProductName': 'Onion (Loose)', 'OriginalPrice': 174.37, 'DiscountPrice': 129.0, 'DiscountPercentage': '26.02%'}


#### **Task 3: Create a list of formatted product descriptions using `map` with a lambda function**

This task demonstrates the `map` function, which applies a given function (in this case, a `lambda`) to all items in an iterable. We'll create a human-readable description for each product.

In [ ]:
# Use map with a lambda function to create formatted product descriptions
product_descriptions = list(map(
    lambda item: f"{item['ProductName']} ({item['qty']} {item['unit_of_measure']}) - {item['Category']}",
    df_as_dict
))

print("First 5 formatted product descriptions:")
for i in range(min(5, len(product_descriptions))):
    print(product_descriptions[i])

First 5 formatted product descriptions:
Onion (Loose) (2.0 kg) - Fruits & Vegetables
Onion (5.0 kg) - Fruits & Vegetables
Onion (Loose) (1.0 kg) - Fruits & Vegetables
Onion (2.0 kg) - Fruits & Vegetables
Onion (Loose) (5.0 kg) - Fruits & Vegetables


#### **Task 4: Build a `GroceryInventoryManager` System using Classes**

This is a challenging task that requires you to apply your knowledge of classes, attributes, methods, and potentially error handling to simulate a basic inventory management system for the grocery warehouse. You'll encapsulate product data and create an object to manage the entire inventory.

**Part 1: The `Product` Class**

Create a Python class named `Product` to represent individual items in the grocery store. Each `Product` object should have the following attributes and methods:

*   **Attributes:**
    *   `product_name` (string)
    *   `category` (string)
    *   `price` (float, original price)
    *   `discount_price` (float, discounted price)
    *   `quantity` (float, current stock level)
    *   `unit_of_measure` (string, e.g., 'kg', 'unit')

*   **Constructor (`__init__`):** Initialize all the above attributes. Ensure `quantity` defaults to `0` if not provided.

*   **Methods:**
    *   `get_current_selling_price(self)`: Returns the `discount_price` if it's less than the `price` and greater than 0; otherwise, returns the `price`. If `discount_price` is 0, it means no discount is applied.
    *   `calculate_discount_percentage(self)`: Calculates and returns the percentage discount based on `price` and `discount_price`. Return 0 if there's no discount or if `price` is 0 to avoid division by zero. Format as a string (e.g., "25.45%").
    *   `update_stock(self, change)`: Adds `change` to the current `quantity`. If `change` is negative, it simulates a sale. Ensure the `quantity` does not drop below zero. Return `True` if successful, `False` otherwise.
    *   `display_info(self)`: Prints a well-formatted string containing all the product's attributes, its current selling price, and its discount percentage.

**Part 2: The `GroceryInventoryManager` Class**

Create a Python class named `GroceryInventoryManager` that will manage a collection of `Product` objects.

*   **Attributes:**
    *   `inventory` (list, to store `Product` objects)

*   **Constructor (`__init__`):** Initialize `inventory` as an empty list.

*   **Methods:**
    *   `add_product(self, product_obj)`: Adds a `Product` object to the `inventory` list. Before adding, check if a product with the same `product_name` already exists. If it does, update its quantity (add to existing stock) instead of adding a duplicate.
    *   `remove_product(self, product_name)`: Removes a `Product` object from the `inventory` by its `product_name`. Handle cases where the product is not found.
    *   `find_product(self, product_name)`: Searches the `inventory` for a product by its `product_name` and returns the `Product` object if found, otherwise returns `None`.
    *   `get_total_inventory_value(self)`: Calculates the sum of (current selling price * quantity) for all products in the inventory. Returns the total value as a float.
    *   `list_products_by_category(self, category)`: Returns a list of `Product` objects that belong to the specified `category`.
    *   `display_all_products(self)`: Iterates through the `inventory` and calls `display_info()` for each `Product` object.

**Part 3: Implementation and Demonstration**

1.  **Initialize Manager:** Create an instance of `GroceryInventoryManager`.
2.  **Populate Inventory:** Use the `df_as_dict` data (from previous Hackathon tasks) to create `Product` objects and add them to your `GroceryInventoryManager` using the `add_product` method. Make sure to handle the `qty` and `unit_of_measure` from `df_as_dict` to initialize the `quantity` and `unit_of_measure` attributes of your `Product` objects.
3.  **Demonstrate Functionality:**
    *   Display the initial state of the entire inventory.
    *   Find and display information about a specific product (e.g., 'Onion (Loose)').
    *   Simulate a sale: "sell" 1.5 kg of 'Onion (Loose)' and then display its updated info.
    *   Simulate a restock: "add" 10.0 kg of 'Onion (Loose)' and display its updated info.
    *   Attempt to sell more of a product than is available and observe the behavior.
    *   Calculate and print the `get_total_inventory_value()` of the entire grocery store.
  

In [1]:
import math

# Part 1: Product Class
class Product:
    def __init__(self, product_name, category, price, discount_price, quantity=0.0, unit_of_measure='unit'):
        self.product_name = product_name
        self.category = category
        self.price = float(price)
        self.discount_price = float(discount_price)
        self.quantity = float(quantity)
        self.unit_of_measure = unit_of_measure

    def get_current_selling_price(self):
        if self.discount_price > 0 and self.discount_price < self.price:
            return self.discount_price
        return self.price

    def calculate_discount_percentage(self):
        if self.price > 0 and self.discount_price > 0 and self.discount_price < self.price:
            percentage = ((self.price - self.discount_price) / self.price) * 100
            return f"{percentage:.2f}%"
        return "0.00%"

    def update_stock(self, change):
        if self.quantity + change >= 0:
            self.quantity += change
            return True
        print(f"Error: Cannot reduce stock below zero for {self.product_name}. Current stock: {self.quantity} {self.unit_of_measure}")
        return False

    def display_info(self):
        print(f"\n--- Product: {self.product_name} ---")
        print(f"  Category: {self.category}")
        print(f"  Original Price: ${self.price:.2f}")
        print(f"  Discount Price: ${self.discount_price:.2f}")
        print(f"  Current Selling Price: ${self.get_current_selling_price():.2f}")
        print(f"  Discount Percentage: {self.calculate_discount_percentage()}")
        print(f"  Stock: {self.quantity:.2f} {self.unit_of_measure}")

# Part 2: GroceryInventoryManager Class
class GroceryInventoryManager:
    def __init__(self):
        self.inventory = []

    def add_product(self, product_obj):
        # Check if product already exists
        existing_product = self.find_product(product_obj.product_name)
        if existing_product:
            existing_product.update_stock(product_obj.quantity)
            print(f"Updated stock for {product_obj.product_name}. New quantity: {existing_product.quantity:.2f} {existing_product.unit_of_measure}")
        else:
            self.inventory.append(product_obj)
            print(f"Added new product: {product_obj.product_name} with quantity {product_obj.quantity:.2f} {product_obj.unit_of_measure}")

    def remove_product(self, product_name):
        original_len = len(self.inventory)
        self.inventory = [p for p in self.inventory if p.product_name != product_name]
        if len(self.inventory) < original_len:
            print(f"Removed product: {product_name}")
            return True
        print(f"Product '{product_name}' not found in inventory.")
        return False

    def find_product(self, product_name):
        for product in self.inventory:
            if product.product_name == product_name:
                return product
        return None

    def get_total_inventory_value(self):
        total_value = 0.0
        for product in self.inventory:
            total_value += product.get_current_selling_price() * product.quantity
        return total_value

    def list_products_by_category(self, category):
        return [p for p in self.inventory if p.category == category]

    def display_all_products(self):
        if not self.inventory:
            print("Inventory is currently empty.")
            return
        print("\n--- Current Grocery Inventory ---")
        for product in self.inventory:
            product.display_info()
        print("-----------------------------------")


In [2]:
# Part 3: Implementation and Demonstration

# 1. Initialize Manager
grocery_manager = GroceryInventoryManager()

# 2. Populate Inventory from df_as_dict
print("\nPopulating inventory...")
for item_data in df_as_dict:
    product = Product(
        product_name=item_data['ProductName'],
        category=item_data['Category'],
        price=item_data['Price'],
        discount_price=item_data['DiscountPrice'],
        quantity=item_data['qty'],
        unit_of_measure=item_data['unit_of_measure']
    )
    grocery_manager.add_product(product)

# 3. Demonstrate Functionality

# Display initial state of the entire inventory
grocery_manager.display_all_products()

# Find and display information about a specific product
print("\n--- Demonstrating find_product and display_info ---")
onion_loose = grocery_manager.find_product('Onion (Loose)')
if onion_loose:
    onion_loose.display_info()
else:
    print("'Onion (Loose)' not found.")

# Simulate a sale: "sell" 1.5 kg of 'Onion (Loose)'
print("\n--- Simulating a sale (1.5 kg of 'Onion (Loose)') ---")
if onion_loose:
    if onion_loose.update_stock(-1.5):
        print("Sale successful.")
    onion_loose.display_info()

# Simulate a restock: "add" 10.0 kg of 'Onion (Loose)'
print("\n--- Simulating a restock (10.0 kg of 'Onion (Loose)') ---")
if onion_loose:
    if onion_loose.update_stock(10.0):
        print("Restock successful.")
    onion_loose.display_info()

# Attempt to sell more of a product than is available
print("\n--- Attempting to over-sell 'Onion (Loose)' (e.g., -100 kg) ---")
if onion_loose:
    onion_loose.update_stock(-100.0)
    onion_loose.display_info()

# Calculate and print the get_total_inventory_value() of the entire grocery store
print("\n--- Total Inventory Value ---")
print(f"Total value of current inventory: ${grocery_manager.get_total_inventory_value():.2f}")


Populating inventory...


NameError: name 'df_as_dict' is not defined